api/app/src/command/medical/document/create-or-update.ts

```diff
@@ -0,0 +1,64 @@
+import { Transaction } from "sequelize";
+import { v4 as uuidv4 } from "uuid";
+import {
+  DocumentReference,
+  DocumentReferenceCreate,
+} from "../../../domain/medical/document-reference";
+import { MedicalDataSource } from "../../../external";
+import { DocumentReferenceModel } from "../../../models/medical/document-reference";
+import { Patient } from "../../../models/medical/patient";
+
+export async function createOrUpdate(
+  patient: Patient,
+  docs: DocumentReferenceCreate[]
+): Promise<DocumentReference[]> {
+  const documents = await Promise.all(
+    docs.map(async doc => {
+      const sequelize = DocumentReferenceModel.sequelize;
+      if (!sequelize) throw new Error("Missing sequelize");
+      let transaction: Transaction | undefined = await sequelize.transaction();
+      try {
+        const existing = await DocumentReferenceModel.findOne({
+          where: {
+            cxId: patient.cxId,
+            patientId: patient.id,
+            source: doc.source,
+            externalId: doc.externalId,
+          },
+          lock: true,
```

Comment:
OOC - is this a row lock or table lock?

Reply:
row lock; were getting OptimisticLockError because the Dash makes two requests to load documents at the same time each time we press the `documents` button.